### PR TITLE
common: support s390 and unknown architectures in spin-wait loop

### PR DESCRIPTION
**Pull request by smithfarm**
Fixes: http://tracker.ceph.com/issues/19492
Signed-off-by: Nathan Cutler <ncutler@suse.com>

### CODE REVIEW

**Comment by smithfarm** [comment]
@badone @tchaikov The jury is still out on whether s390 has such an assembler instruction. This is one way the issue could be dealt with - are we sure we want to break builds for architectures not recognized in this `#if` block, though? Would it make sense to put the old code into the `#else` ?

**Comment by tchaikov** [comment]
> are we sure we want to break builds for architectures not recognized in this #if block, though? 

i am not sure now. maybe we should `#else` them.

**Comment by smithfarm** [comment]
@kefu So, like this?

File: src/common/simple_spin.cc [hunk 1]
```diff
@@ -18,6 +18,11 @@
 #include <stdint.h>
 #include <pthread.h>
 
+#if !defined(__i386__) && !defined(__x86_64__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc__) && !defined(__ppc__)
```
Comment (tchaikov) [inline src/common/simple_spin.cc]: i feel guilty at seeing this. but it works!

**Comment by smithfarm** [comment]
One colleague suggested the detail namespace idiom, so I added a commit for that. Another colleague wrote:

> I don't know s390, but looks like in kernel we have just a simple compiler barrier for s390 implementation of cpu_relax() (which is done by "asm pause" on x86).
> 
> The compiler barrier is:
> 
> #define barrier() __asm__ __volatile__("": : :"memory")

So I added a commit for that as well. Please re-review.

**Comment by smithfarm** [comment]
Pushed `wip-19492` to ceph-ci.

**Comment by smithfarm** [comment]
Builds fine on shaman, also builds successfully on all four architectures (x86_64, aarch64, ppc64le, and s390x) in OBS.

**Comment by badone** [comment]
I'd be interested to see the assembly produced for that on s390(x) (even though I know little about the arch). On x86_64 at least it appears to produce no instructions, https://godbolt.org/g/i6vrMP If that is what we are using in the kernel (I looked originally and this is what I found also) I guess it's the best we can do. I puzzle over what sort of DAS s390(x) ceph installations are using, but I guess that's a discussion for another forum?

File: src/common/simple_spin.cc [hunk 2]
```diff
@@ -18,25 +18,42 @@
...
+      *foo = (*foo * 33) + 17;
+    }
 #endif
+}
+
+} // namespace detail
+
+void simple_spin_lock(simple_spinlock_t *lock)
+{
+  while(1) {
```
Comment (tchaikov) [inline src/common/simple_spin.cc]: could take this chance to add a space after `while`.

File: src/common/simple_spin.cc [hunk 3]
```diff
@@ -18,25 +18,42 @@
 #include <stdint.h>
 #include <pthread.h>
 
-void simple_spin_lock(simple_spinlock_t *lock)
+namespace detail {
```
Comment (tchaikov) [inline src/common/simple_spin.cc]: i don't think `detail` namespace would help here, but an anonymous namespace does. and would be better if we can rename `delay` to something like `cpu_relax` or just `pause`.

**Comment by smithfarm** [comment]
@badone The purpose of the s390 build is to enable mainframe clients to access external Ceph clusters over the network.

File: src/common/simple_spin.cc (hunk 3 above)
Comment (smithfarm) [inline src/common/simple_spin.cc]: ok, done

File: src/common/simple_spin.cc (hunk 2 above)
Comment (smithfarm) [inline src/common/simple_spin.cc]: done

**Comment by smithfarm** [comment]
@badone Regarding `"": : :"memory"` I found this: https://en.wikipedia.org/wiki/Memory_ordering#Compile-time_memory_barrier_implementation

**Comment by smithfarm** [comment]
As agreed on IRC, pushed latest version to ceph-ci and will run it through `rados/thrash` for validation.

**Comment by smithfarm** [comment]
`teuthology-suite -k distro --priority 101 --suite rados/thrash --subset $(expr $RANDOM % 2000)/2000 --email ncutler@suse.com --ceph wip-19492 --machine-type smithi`

**waiting for build**

**Pull request merged**